src/mpid/ch4/shm/xpmem/xpmem_seg.h

```diff
@@ -0,0 +1,644 @@
+                if (lct_height + 1 == rheight) {
+                    /* Insert into left child of right child */
+                    /* Do right-left rotation */
+                    mpi_errno = MPIDI_XPMEM_avl_do_right_left_rotation(sub_root, ct_node);
+                    if (mpi_errno != MPI_SUCCESS)
+                        MPIR_ERR_POP(mpi_errno);
+                } else {
+                    /* Insert into right child of right child */
+                    /* Do left rotation */
+                    mpi_errno = MPIDI_XPMEM_avl_do_left_rotation(sub_root, ct_node);
+                    if (mpi_errno != MPI_SUCCESS)
+                        MPIR_ERR_POP(mpi_errno);
+                }
+            }
+
+            if (!MPIDI_XPMEM_AVL_STACK_EMPTY(node_stack)) {
+                MPIDI_XPMEM_AVL_STACK_POP(node_stack, sub_root);
+                goto stack_recovery;
+            } else
+                break;
+
+        } while (1);
+
+        while (tree->root->parent != NULL)
+            tree->root = tree->root->parent;
+    }
+
+  fn_exit:
+    MPID_THREAD_CS_EXIT(VCI, tree->lock);
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_XPMEM_SEGTREE_DO_SEARCH_AND_INSERT_SAFE);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+/****************************************/
+/* Segment cache public routines        */
+/****************************************/
+
+/* Initialize an empty tree for segment cache.
+ * It should be called only once for a AVL tree at MPI init.*/
+#undef FUNCNAME
+#define FUNCNAME MPIDI_XPMEM_segtree_init
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_segtree_init(MPIDI_XPMEM_segtree_t * tree)
+{
+    int mpi_errno = MPI_SUCCESS, ret;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_SEGTREE_INIT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_SEGTREE_INIT);
+
+    tree->root = NULL;
+    tree->tree_size = 0;
+    MPID_Thread_mutex_create(&tree->lock, &ret);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_XPMEM_SEGTREE_INIT);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+/* Deletes all registered segments in the segment cache.
+ * It detaches and frees each segment. */
+#undef FUNCNAME
+#define FUNCNAME MPIDI_XPMEM_segtree_delete_all
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_segtree_delete_all(MPIDI_XPMEM_segtree_t * tree)
+{
+    uint8_t direction;
+    int mpi_errno = MPI_SUCCESS, ret;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_SEGTREE_DELETE_ALL);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_SEGTREE_DELETE_ALL);
+
+    MPIDI_XPMEM_AVL_DECLARE_STACK(node_stack, MPIDI_XPMEM_seg_t *, MPIDI_XPMEM_AVL_STACK_SIZE);
+    MPIDI_XPMEM_AVL_DECLARE_STACK(direction_stack, uint8_t, MPIDI_XPMEM_AVL_STACK_SIZE);
+
+    if (tree->tree_size) {
+        MPIDI_XPMEM_seg_t *node = tree->root;
+        while (1) {
```

Comment:
I also put the comments here:

> /* this delete function will traverse all the segments stored in AVL
             * tree, and this infinite loop is used to avoid recursive call of 
             * funtion. It exits when all segments have been deletes (also means
             * stack is empty). */